akka-http-core/src/main/scala/akka/http/impl/engine/client/OutgoingConnectionBlueprint.scala

```diff
@@ -38,62 +42,75 @@ private[http] object OutgoingConnectionBlueprint {
     responseOut  |  responsePrep  |  Response  |<---+                                       |
     <------------+----------------|  Parsing   |                                            |
                                   |  Merge     |<------------------------------------------ V
                                   +------------+
   */
-  def apply(hostHeader: Host,
+  def apply(hostHeader: headers.Host,
             settings: ClientConnectionSettings,
             log: LoggingAdapter): Http.ClientLayer = {
     import settings._
 
-    // the initial header parser we initially use for every connection,
-    // will not be mutated, all "shared copy" parsers copy on first-write into the header cache
-    val rootParser = new HttpResponseParser(parserSettings, HttpHeaderParser(parserSettings) { info ⇒
-      if (parserSettings.illegalHeaderWarnings)
-        logParsingError(info withSummaryPrepended "Illegal response header", log, parserSettings.errorLoggingVerbosity)
-    })
+    val core = BidiFlow.fromGraph(GraphDSL.create() { implicit b ⇒
+      import GraphDSL.Implicits._
 
-    val requestRendererFactory = new HttpRequestRendererFactory(userAgentHeader, requestHeaderSizeHint, log)
+      val renderingContextCreation = b.add {
+        Flow[HttpRequest] map { request ⇒
+          val sendEntityTrigger =
+            request.headers collectFirst { case headers.Expect.`100-continue` ⇒ Promise[Unit]().future }
+          RequestRenderingContext(request, hostHeader, sendEntityTrigger)
+        }
+      }
 
-    val requestRendering: Flow[HttpRequest, ByteString, NotUsed] = Flow[HttpRequest]
-      .map(RequestRenderingContext(_, hostHeader))
-      .via(Flow[RequestRenderingContext].flatMapConcat(requestRendererFactory.renderToSource).named("renderer"))
+      val bypassFanout = b.add(Broadcast[RequestRenderingContext](2, eagerCancel = true))
 
-    val methodBypass = Flow[HttpRequest].map(_.method)
+      val terminationMerge = b.add(TerminationMerge)
 
-    import ParserOutput._
-    val responsePrep = Flow[List[ResponseOutput]]
-      .mapConcat(conforms)
-      .via(new PrepareResponse(parserSettings))
+      val requestRendering: Flow[RequestRenderingContext, ByteString, NotUsed] = {
+        val requestRendererFactory = new HttpRequestRendererFactory(userAgentHeader, requestHeaderSizeHint, log)
+        Flow[RequestRenderingContext].flatMapConcat(requestRendererFactory.renderToSource).named("renderer")
+      }
 
-    val core = BidiFlow.fromGraph(GraphDSL.create() { implicit b ⇒
-      import GraphDSL.Implicits._
-      val methodBypassFanout = b.add(Broadcast[HttpRequest](2, eagerCancel = true))
-      val responseParsingMerge = b.add(new ResponseParsingMerge(rootParser))
+      val bypass = Flow[RequestRenderingContext] map { ctx ⇒
+        HttpResponseParser.ResponseContext(ctx.request.method, ctx.sendEntityTrigger.map(_.asInstanceOf[Promise[Unit]]))
```

Comment:
If Future were to change in 2.11.9, it would break all released artifacts
that relied on such a hack. Silently. :-)

-- 
Cheers,
√
On Mar 10, 2016 9:35 PM, "Mathias" <notifications@github.com> wrote:

> In
> akka-http-core/src/main/scala/akka/http/impl/engine/client/OutgoingConnectionBlueprint.scala
> <https://github.com/akka/akka/pull/19943#discussion_r55745254>:
>
> >
> > -    val core = BidiFlow.fromGraph(GraphDSL.create() { implicit b ⇒
> > -      import GraphDSL.Implicits._
> > -      val methodBypassFanout = b.add(Broadcast[HttpRequest](2, eagerCancel = true))
> > -      val responseParsingMerge = b.add(new ResponseParsingMerge(rootParser))
> > +      val bypass = Flow[RequestRenderingContext] map { ctx ⇒
> > +        HttpResponseParser.ResponseContext(ctx.request.method, ctx.sendEntityTrigger.map(_.asInstanceOf[Promise[Unit]]))
>
> Yeah, it's a bit of a hack.
> But any change in the underlying Future implementation will break several
> tests, so it won't be silent.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/akka/akka/pull/19943/files#r55745254>.
>